cipher/cipher.go

```diff
@@ -0,0 +1,186 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package cipher
+
+import (
+	"crypto/aes"
+	"crypto/cipher"
+	"fmt"
+	"sync"
+)
+
+type Mode string
+type Padding string
+
+const (
+	ModeEcb = Mode("ECB")
+)
+
+const (
+	PaddingPKCS7 = Padding("PKCS7Padding")
+	PaddingPKCS5 = Padding("PKCS5Padding")
+)
+
+var (
+	supportedModes = map[Mode]bool{
+		ModeEcb: true,
+	}
+	supportedPadding = map[Padding]bool{
+		PaddingPKCS7: true,
+		PaddingPKCS5: true,
+	}
+)
+
+func (p Padding) padding(input []byte) (output []byte) {
+	switch p {
+	case PaddingPKCS5:
+		fallthrough
+	case PaddingPKCS7:
+		//PKCS7Padding
+		numPad := 16 - (len(input) % 16)
+		output = make([]byte, len(input)+numPad)
+		for i := copy(output, []byte(input)); i < len(output); i++ {
+			output[i] = byte(numPad)
+		}
+	}
+	return
+}
+
+func (p Padding) strip(input []byte) (output []byte) {
+	switch p {
+	case PaddingPKCS5:
+		fallthrough
+	case PaddingPKCS7:
+		//remove PKCS7Padding
+		numPad := int(input[len(input)-1])
+		output = input[:(len(input) - numPad)]
+	}
+	return
+}
+
+// Create a new AesCipher object with the specified encryption mode and padding algorithm.
+func CreateAesCipher(key []byte) (*AesCipher, error) {
+	a := &AesCipher{
+		mutex: &sync.RWMutex{},
+	}
+	if err := a.SetKey(key); err != nil {
+		return nil, err
+	}
+	return a, nil
+}
+
+// An object to perform AES encryption/decryption.
+type AesCipher struct {
+	key   []byte
+	block cipher.Block
+	mutex *sync.RWMutex
```

Comment:
I don't think we should do locking around code that does decryption, there is no need to share the AesCipher object among threads

Reply:
An `AesCipher` instance is shared among all request handlers, which happen concurrently. Yes you're right, if we keep the `key` immutable we can get rid of locking.